Sources/HummingbirdJobs/JobQueueHandler.swift

```diff
@@ -12,89 +12,94 @@
 //
 //===----------------------------------------------------------------------===//
 
+import AsyncAlgorithms
 import Hummingbird
 import Logging
+import ServiceLifecycle
 
 /// Object handling a single job queue
-public final class HBJobQueueHandler {
-    public init(queue: HBJobQueue, numWorkers: Int, eventLoopGroup: EventLoopGroup, logger: Logger) {
+public final class HBJobQueueHandler<Queue: HBJobQueue>: Service {
+    public init(queue: Queue, numWorkers: Int, logger: Logger) {
         self.queue = queue
-        self.workers = (0..<numWorkers).map { _ in
-            HBJobQueueWorker(queue: queue, eventLoop: eventLoopGroup.next(), logger: logger)
-        }
-        self.eventLoop = eventLoopGroup.next()
+        self.numWorkers = numWorkers
+        self.logger = logger
     }
 
     /// Push Job onto queue
     /// - Returns: Queued job information
-    @discardableResult public func enqueue(_ job: HBJob, on eventLoop: EventLoop) -> EventLoopFuture<JobIdentifier> {
-        self.queue.enqueue(job, on: eventLoop)
+    @discardableResult public func enqueue(_ job: HBJob) async throws -> JobIdentifier {
+        try await self.queue.push(job)
     }
 
-    /// Start queue workers
-    public func start() {
-        self.queue.onInit(on: self.eventLoop).whenComplete { _ in
-            self.workers.forEach {
-                $0.start()
+    public func run() async throws {
+        try await withGracefulShutdownHandler {
+            try await withThrowingTaskGroup(of: Void.self) { group in
```

Comment:
Shouldn't this be a discarding task group?

Reply:
A discarding task group doesn't give you any info on when a child task has finished. So I have no way to control how many tasks are running. I think a normal task group is ok as long as you are calling next() as that'll drop finished tasks off the list. I will need to check that.

Reply:
Going to merge this and revisit the usage to withTaskGroup